src/ziggurat/messaging/producer.clj

```diff
@@ -79,29 +80,55 @@
           (handle-network-exception e message-payload))
         (catch IOException e
           (handle-network-exception e message-payload))
+        (catch TimeoutException e
+          (handle-network-exception e message-payload))
         (catch Exception e
           (log/error e "Exception was encountered while publishing to RabbitMQ")
           (metrics/increment-count ["rabbitmq" "publish"] "exception" {:topic-entity (name (:topic-entity message-payload))})
-          false)
+          :retry-with-counter)
         (finally (return-to-pool cpool/channel-pool ch))))
     (catch Exception e
-      (log/error e "Exception occurred while borrowing a channel from the pool")
+      (log/error e "Exception was encountered while borrowing a channel from the pool")
       (metrics/increment-count ["rabbitmq" "publish" "channel_borrow"] {:topic-entity (name (:topic-entity message-payload))})
-      false)))
+      :retry-with-counter)))
+
+(defn- publish-retry-config []
+  (-> (ziggurat-config) :rabbit-mq-connection :publish-retry))
+
+(defn- non-recoverable-exception-config []
+  (:non-recoverable-exception (publish-retry-config)))
 
 (defn publish
+  "This is meant for publishing to rabbitmq.
+  * Checks if the pool is alive - We do this so that publish does not happen after the channel pool state is stopped.
+  * publish-internal returns multiple states
+    * :success - Message has been successfully produced to rabbitmq
+    * :retry - A retryable exception was encountered and message will be retried until it is successfully published.
+    * :retry-with-counter - A non recoverable exception is encountered, but the message will be retried for a few times. defined by the counter
+      { :rabbit-mq-connection { :publish-retry { :non-recoverable-exception {:count}}}}}"
   ([exchange message-payload]
    (publish exchange message-payload nil))
   ([exchange message-payload expiration]
-   (when (publish-internal exchange message-payload expiration)
-     (Thread/sleep 5000)
-     (log/info "Retrying publishing the message to " exchange)
-     (recur exchange message-payload expiration))))
+   (publish exchange message-payload expiration (:count (non-recoverable-exception-config))))
+  ([exchange message-payload expiration counter]
+   (when (is-pool-alive? cpool/channel-pool)
+     (let [result (publish-internal exchange message-payload expiration)]
+       (cond
+         (= result :success) nil
+         (= result :retry) (do
+                             (Thread/sleep (:sleep (publish-retry-config)))
+                             (log/info "Retrying publishing the message to " exchange)
+                             (recur exchange message-payload expiration counter))
+         (= result :retry-with-counter) (if (and (:enabled (non-recoverable-exception-config)) (pos? counter))
+                                          (do
+                                            (Thread/sleep (:sleep (non-recoverable-exception-config)))
```

Comment:
Instead of calling this config `:sleep` can we change it to `:back-off-ms` or something along a similar line